app/features/build_runner/remote_runner.rb

```diff
@@ -20,187 +28,193 @@ class RemoteRunner
-      elsif (@build_state.state == :BROKEN || @build_state.state == :REJECTED) && !response.error
-        raise "Expecting InvocationResponses containing the error details after :BROKEN or :REJECTED"
-      elsif @build_state.state == :FINISHING && !response.artifact
-        raise "Expecting InvocationResponses containing artifacts while :FINISHING"
-      elsif @build_state.state == :SUCCEEDED || @build_state.state == :FAILED
-        raise "No further InvocationResponse messages expected after :SUCCEEDED or :FAILED"
+    def handle_log(log)
+      logger.debug("handle log: #{log.inspect}")
+      publish_to_all(log: log.to_h)
+    end
+
+    def handle_error(error)
+      logger.debug("handle error: #{error}")
+      publish_to_all(error: error.to_h)
+    end
+
+    # handle artifact upload. Do not publish this to the subscribers.
+    def handle_artifact(artifact)
+      logger.debug("handle artifact: #{artifact.filename}")
+
+      artifact_path = File.join(artifacts_path, artifact.filename)
+
+      unless current_build.artifacts.any? { |a| a.reference == artifact_path }
+        current_build.artifacts << Artifact.new(
+          type: artifact.filename,
+          reference: artifact_path,
+          provider: project.artifact_provider
+        )
       end
+
+      # open (or create if it doesn't exist) the file for appending with binary data.
+      File.open(artifact_path, "ab+") { |f| f.write(artifact.chunk) }
     end
 
-    def process_agent_response(response:)
-      if response.state != :PENDING
-        @build_state.trigger(response.state)
-      else
-        case @build_state.state
-        when :REJECTED
-          # Agent is busy. How do we handle this?
-          emit_error_response(response.error)
-        when :RUNNING
-          emit_new_row(
-            type: response.log.level,
-            message: response.log.message,
-            time: Time.now
-          )
-        when :FINISHING
-          artifact_path = File.join(@artifacts_path, response.artifact.filename)
-          artifact = current_build.artifacts.find { |a| a.reference == artifact_path }
-          unless artifact
-            current_build.artifacts << Artifact.new(
-              type: response.artifact.filename,
-              reference: artifact_path,
-              provider: @project.artifact_provider
-            )
-          end
-          File.open(artifact_path, "a") { |f| f.write(response.artifact.chunk) }
-        when :BROKEN
-          emit_error_response(response.error)
-        end
+    # subscribe will all send all historic data to the subscriber.
+    # the yielded object will be an InvocationResponse object
+    def subscribe(&block)
+      logger.debug("subscribing listener to topic `#{topic_name}`")
+      subscriber = FastlaneCI::Notifications.subscribe(topic_name, &block)
+
+      replay_history_to(subscriber)
+
+      # make sure we call completion blocks if a subscriber joins after we have completed.
```

Comment:
Just wondering if we shouldn't block the subscribe once we are done. What are the cases when one would subscribe after completion and also, what is the lifetime of the runner - do we keep it around after is done?

Reply:
That makes _a lot_ of sense 😄

The lifetime is not defined. We don't explicitly remove the runner, so it lives until the process is restarted. Perhaps we should explicitly removed it.